src/platforms/common/stm32/serialno.c

```diff
@@ -64,3 +72,32 @@ void read_serial_number(void)
 #endif
 	serial_no[DFU_SERIAL_LENGTH - 1] = '\0';
 }
+
+static char *utoa_upper(unsigned value, char *str, int base)
+{
+	const char digits[] = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
```

Comment:
This must be marked `static` to prevent its pointer being copied to stack every run.

Reply:
Yes, I agree, codegen looks horrible (entire array gets copied via mass ldmia!/stmia! and 4 regs from `.rodata`). Updated to `static`. Do you have any ideas on making it `utoa_lut` and feeding it different `digits[]` arrays to reuse the same algorithm?

Reply:
Unfortunately no particular good ideas spring to mind immediately, but if any do come to mind we'll let you know - for now, we'll resolve this thread as this looks good.